osu.Game/Screens/Select/Carousel/CarouselBeatmapSet.cs

```diff
@@ -61,50 +61,79 @@ public override int CompareTo(FilterCriteria criteria, CarouselItem other)
+                    break;
 
                 case SortMode.DateAdded:
-                    return otherSet.BeatmapSet.DateAdded.CompareTo(BeatmapSet.DateAdded);
+                    comparison = otherSet.BeatmapSet.DateAdded.CompareTo(BeatmapSet.DateAdded);
+                    break;
 
                 case SortMode.DateRanked:
                     // Beatmaps which have no ranked date should already be filtered away in this mode.
                     if (BeatmapSet.DateRanked == null || otherSet.BeatmapSet.DateRanked == null)
-                        return 0;
+                        break;
 
-                    return otherSet.BeatmapSet.DateRanked.Value.CompareTo(BeatmapSet.DateRanked.Value);
+                    comparison = otherSet.BeatmapSet.DateRanked.Value.CompareTo(BeatmapSet.DateRanked.Value);
+                    break;
 
                 case SortMode.LastPlayed:
-                    return -compareUsingAggregateMax(otherSet, b => (b.LastPlayed ?? DateTimeOffset.MinValue).ToUnixTimeSeconds());
+                    comparison = -compareUsingAggregateMax(otherSet, b => (b.LastPlayed ?? DateTimeOffset.MinValue).ToUnixTimeSeconds());
+                    break;
 
                 case SortMode.BPM:
-                    return compareUsingAggregateMax(otherSet, b => b.BPM);
+                    comparison = compareUsingAggregateMax(otherSet, b => b.BPM);
+                    break;
 
                 case SortMode.Length:
-                    return compareUsingAggregateMax(otherSet, b => b.Length);
+                    comparison = compareUsingAggregateMax(otherSet, b => b.Length);
+                    break;
 
                 case SortMode.Difficulty:
-                    return compareUsingAggregateMax(otherSet, b => b.StarRating);
+                    comparison = compareUsingAggregateMax(otherSet, b => b.StarRating);
+                    break;
 
                 case SortMode.DateSubmitted:
                     // Beatmaps which have no submitted date should already be filtered away in this mode.
                     if (BeatmapSet.DateSubmitted == null || otherSet.BeatmapSet.DateSubmitted == null)
-                        return 0;
+                        break;
 
-                    return otherSet.BeatmapSet.DateSubmitted.Value.CompareTo(BeatmapSet.DateSubmitted.Value);
+                    comparison = otherSet.BeatmapSet.DateSubmitted.Value.CompareTo(BeatmapSet.DateSubmitted.Value);
+                    break;
             }
+
+            if (comparison != 0) return comparison;
+
+            // If the initial sort could not differentiate, attempt to use DateAdded and OnlineID to order sets in a stable fashion.
+            // This directionality is a touch arbitrary as while DateAdded puts newer beatmaps first, the OnlineID fallback puts lower IDs first.
+            // Can potentially be changed in the future if users actually notice / have preference, but keeping it this way matches historical tests.
```

Comment:
I'm *pretty* sure it only matches historical tests because of a test implementation detail, namely this:

https://github.com/ppy/osu/blob/53d968499eba134545ec8d6e5dc4cd84346b4f21/osu.Game.Tests/Resources/TestResources.cs#L91

Each call to `TestResources.CreateTestBeatmapSetInfo()` would choose the next free set ID and increment a static counter, so essentially what the test was exercising is that the order of insertion is not disturbed. Based on that, I'm not super sure we need to be making special exceptions for what is essentially a test oddity; I'd probably remove the `OnlineID` behaviour, as right now it is adding a lot of extra weirdness and complication. This is probably most apparent in the fact that the comparison exists at all and the switched directionality has to be explained at all..

Reply:
Fair, have made the changes required.